File: 2.两数相加.cs
/*
 * @lc app=leetcode.cn id=2 lang=csharp
 *
 * [2] 两数相加
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
        public ListNode AddTwoNumbers(ListNode l1, ListNode l2)
        {
            ListNode l_add = new ListNode(0);
            l_add = resarch(l1, l2, l_add.next = new ListNode(0),0);
            return l_add;
        }

        public ListNode resarch(ListNode l1, ListNode l2, ListNode l_normal,int count)
        {
            l1 = l1 == null ? new ListNode(0) : l1;
            l2 = l2 == null ? new ListNode(0) : l2;
            l_normal.val = l1.val + l2.val + count;
            count = 0;
            if (l_normal.val >= 10)
            {
                l_normal.val = l_normal.val%10;
                count++;
            }
            if (count!=0||(l1.next != null || l2.next != null))
            {
                resarch(l1.next, l2.next, l_normal.next = new ListNode(0), count);
            }
            return l_normal;
        }
}

File: 2.两数相加.ts
/*
 * @lc app=leetcode.cn id=2 lang=typescript
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 * 1.0  遍历 当有任何乙方存在时 都继续 生成新的队列 保留进制数
 * 基准条件  溢出为0 数值A为空 数值B为空
 */
function addTwoNumbers(
  l1: ListNode | null,
  l2: ListNode | null
): ListNode | null {
  let rootListNode = new ListNode();
  let currentNode = null;
  let isRoot = true;
  let overflow = 0;
  while (l1 || l2 || overflow) {
    if (isRoot) {
      currentNode = rootListNode;
      isRoot = false;
    } else {
      currentNode.next = new ListNode();
      currentNode = currentNode.next;
    }
    if (!l1) l1 = new ListNode(0);
    if (!l2) l2 = new ListNode(0);
    let val = l1.val + l2.val + overflow;
    overflow = Math.floor(val / 10);
    currentNode.val = overflow === 1 ? val % 10 : val;
    l1 = l1.next;
    l2 = l2.next;
  }

  return rootListNode;
}

// function recursive(
//   l1: ListNode | null,
//   l2: ListNode | null,
//   currentNode: ListNode,
//   overflow: number = 0
// ) {
//   if (!l1) l1 = new ListNode(0);
//   if (!l2) l2 = new ListNode(0);
//   let val = l1.val + l2.val + overflow;
//   overflow = Math.floor(val / 10);
//   currentNode.val = overflow === 1 ? val % 10 : val;

//   /**基线条件  溢出为0 数值A为空 数值B为空*/
//   if (!l1.next && !l2.next && overflow === 0) {
//     return currentNode;
//   }

//   let childNode = new ListNode();
//   currentNode.next = childNode;
//   recursive(l1.next, l2.next, childNode, overflow);
// }

// function addTwoNumbers(
//   l1: ListNode | null,
//   l2: ListNode | null
// ): ListNode | null {
//   let rootListNode = new ListNode();
//   recursive(l1, l2, rootListNode, 0);
//   return rootListNode;
// }
// @lc code=end
